05-input-output.Rmd

```diff
@@ -264,6 +263,10 @@ The results of this simple disk usage benchmark show that saving data in a compr
 
 The results show that the relative *size* of different formats is not a reliable predictor of data read and write times. This is due to the computational overheads of compression. Although the binary `.feather` format did not perform well in terms of read and write times, the function `read_feather` is *faster*  than R's native `.Rds`  format, for the datasets used in the benchmark. `write_feather` is also faster than `saveRDS` for all but the largest dataset. In all cases, `read.csv` and `write.csv` is several times slower than the binary formats and this relative slowness worsens with increasing dataset size. In the next section we explore the performance of alternatives to these base R functions for reading and writing plain text data files.
 
+### Protocol Buffers
+
+Google's [Protocol Buffers](https://developers.google.com/protocol-buffers/) offers a portable, efficient and scalable solution to binary data storage. A recent package, **RProtoBuf**, provides an R interface. This approach is not covered in this book, as it is new, advanced and not (at the time of writing) widely used in the R community. The approach is discussed in detail in a [paper](https://www.jstatsoft.org/article/view/v071i02) on the subject, which also provides an excellent overview of the issues associated with different file formats [@JSSv071i02].
```

Comment:
Thanks. I'm on Ubuntu 12.04 and the error message was:
```
onfigure: error: ERROR: ProtoBuf headers required; use '-Iincludedir' in CXXFLAGS for unusual locations.
ERROR: configuration failed for package ‘RProtoBuf’
* removing ‘/home/robin/R/x86_64-pc-linux-gnu-library/3.3/RProtoBuf’
Warning in install.packages :
  installation of package ‘RProtoBuf’ had non-zero exit status
```

I'll check out the appendix, thanks for the pointer - will try to figure it out.

Seems really promising for increasing the portability of binary files generated by R.